src/lm/const-arpa-lm.cc

```diff
@@ -937,8 +944,10 @@ void ConstArpaLm::WriteArpaRecurse(int32* lm_state,
   // Inserts the current LmState to <output>.
   ArpaLine arpa_line;
   arpa_line.words = seq;
-  arpa_line.logprob = *reinterpret_cast<float*>(lm_state);
-  arpa_line.backoff_logprob = *reinterpret_cast<float*>(lm_state + 1);
+  Int32AndFloat logprob_i(*lm_state);
+  arpa_line.logprob = logprob_i.f;
```

Comment:
lm_state has type of int32*. Or perhaps you mean casting lm_state to
Int32AndFloat using something like ((union Int32AndFloat *)lm_state)->f?
That may still cause strict aliasing problem?

Guoguo

On Mon, Oct 5, 2015 at 8:00 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/lm/const-arpa-lm.cc
> <https://github.com/kaldi-asr/kaldi/pull/211#discussion_r41213113>:
>
> > @@ -937,8 +944,10 @@ void ConstArpaLm::WriteArpaRecurse(int32* lm_state,
> >    // Inserts the current LmState to <output>.
> >    ArpaLine arpa_line;
> >    arpa_line.words = seq;
> > -  arpa_line.logprob = *reinterpret_cast<float*>(lm_state);
> > -  arpa_line.backoff_logprob = *reinterpret_cast<float*>(lm_state + 1);
> > +  Int32AndFloat logprob_i(*lm_state);
> > +  arpa_line.logprob = logprob_i.f;
>
> I don't see why a temporary is needed... just use lm_state->f.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/211/files#r41213113>.
>